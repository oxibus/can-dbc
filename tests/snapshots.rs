#![cfg(all(feature = "serde", feature = "encodings"))]

use nom::Err;
use std::borrow::Cow;
use std::fs;
use std::path::{Path, PathBuf};

use can_dbc::{decode_cp1252, Dbc, Error};
use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use test_each_file::test_each_path;

struct TestConfig {
    test_root: &'static str,
    snapshot_suffix: &'static str,
    use_cp1251: bool,
    create_snapshot: bool,
}

static TEST_DIRS: &[TestConfig] = &[
    TestConfig {
        test_root: "shared-test-files",
        snapshot_suffix: "",
        use_cp1251: true,
        create_snapshot: true,
    },
    TestConfig {
        test_root: "opendbc/opendbc/dbc",
        snapshot_suffix: "opendbc",
        use_cp1251: false,
        create_snapshot: false,
    },
];

test_each_path! { for ["dbc"] in "./tests/fixtures/opendbc/opendbc/dbc" as dbc => parse_one_file }
test_each_path! { for ["dbc"] in "./tests/fixtures/shared-test-files" as shared => parse_one_file }
// upper case extension
test_each_path! { for ["DBC"] in "./tests/fixtures/shared-test-files" as shared2 => parse_one_file }

/// Get `test root`, `snapshot name suffix`, `use cp1251`, `create snapshot` for the given path
fn get_test_info(path: &Path) -> Option<(PathBuf, &'static TestConfig)> {
    if !path
        .extension()
        .unwrap_or_default()
        .eq_ignore_ascii_case("dbc")
    {
        return None;
    }
    let path_str = path.to_str().unwrap();
    let parent = path.parent().unwrap();
    for item in TEST_DIRS {
        // Ensure slashes are there for easier matching
        let test_root = format!("/{}/", item.test_root);
        let mut path_dir = parent.to_str().unwrap().to_string();
        if !path_dir.ends_with('/') {
            path_dir.push('/');
        }
        if let Some(pos) = path_dir.find(&test_root) {
            let parent = PathBuf::from("snapshots")
                .join(item.snapshot_suffix)
                .join(&path_dir[pos + test_root.len()..]);
            return Some((parent, item));
        }
    }
    panic!("Unknown test directory: {path_str}");
}

/// Test parsing all DBC files
#[test]
fn test_if_submodules_are_present() {
    for test in TEST_DIRS {
        let dir = Path::new("./tests/fixtures").join(test.test_root);
        fs::read_dir(&dir).unwrap_or_else(|e| {
            let dir_display = dir.display();
            panic!(
                "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
            )
        });
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file([path]: [&Path; 1]) {
    let Some((
        snapshot_path,
        &TestConfig {
            use_cp1251,
            create_snapshot,
            ..
        },
    )) = get_test_info(path)
    else {
        return;
    };

    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if use_cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    report_test_result(
        Dbc::try_from(buffer.as_ref()),
        snapshot_path,
        file_name,
        create_snapshot,
    );
}

fn report_test_result(
    result: Result<Dbc, Error>,
    snapshot_path: PathBuf,
    file_name: String,
    do_snapshot: bool,
) {
    let do_snapshot = std::env::var("SKIP_INSTA").is_err()
        && (do_snapshot || std::env::var("FORCE_INSTA").is_ok());

    with_settings! {
        {
            omit_expression => true,
            snapshot_path => snapshot_path,
            prepend_module_to_snapshot => false
        },
        {
            match result {
                Ok(result) => {
                    if do_snapshot {
                        assert_yaml_snapshot!(file_name, result);
                    } // else ignore successful results when not snapshotting
                }
                Err(e) => {
                    let file_name = format!("!error___{file_name}");
                    assert_debug_snapshot!(file_name, err_to_str(&e));
                }
            }
        }
    }
}

fn err_to_str<'a>(err: &'a Error) -> &'a str {
    match err {
        Error::Incomplete(_, _) => "incomplete",
        Error::Nom(e) => match e {
            Err::Incomplete(_) => "nom-incomplete",
            Err::Error(_) => "nom-error",
            Err::Failure(_) => "nom-failure",
        },
        Error::MultipleMultiplexors => "multiple multiplexors",
    }
}
