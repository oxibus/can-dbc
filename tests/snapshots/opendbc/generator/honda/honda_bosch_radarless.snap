---
source: tests/snapshots.rs
---
version: ""
bit_timing: ~
messages:
  - id:
      Standard: 456
    name: ACC_CONTROL
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: ACCEL_COMMAND
        multiplexer_indicator: Plain
        start_bit: 7
        size: 12
        byte_order: BigEndian
        value_type: Signed
        factor: 0.01
        offset: 0
        min: 0
        max: 0
        unit: m/s^2
        receivers:
          - XXX
      - name: IDLESTOP_ALLOW
        multiplexer_indicator: Plain
        start_bit: 8
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: STANDSTILL
        multiplexer_indicator: Plain
        start_bit: 9
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: CONTROL_ON
        multiplexer_indicator: Plain
        start_bit: 10
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: BOH
        multiplexer_indicator: Plain
        start_bit: 23
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: AEB_STATUS
        multiplexer_indicator: Plain
        start_bit: 33
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 7
        unit: ""
        receivers:
          - XXX
      - name: AEB_BRAKING
        multiplexer_indicator: Plain
        start_bit: 47
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: AEB_PREPARE
        multiplexer_indicator: Plain
        start_bit: 43
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: FCW_1
        multiplexer_indicator: Plain
        start_bit: 24
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: FCW_2
        multiplexer_indicator: Plain
        start_bit: 54
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 59
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 61
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 467
    name: CRUISE_FAULT_STATUS
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CRUISE_FAULT
        multiplexer_indicator: Plain
        start_bit: 3
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 59
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 61
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 477
    name: GEARBOX_ALT_2
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: GEAR_MT
        multiplexer_indicator: Plain
        start_bit: 39
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 59
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 61
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 495
    name: SPEED_LIMIT_DASH_DISPLAY
    size: 8
    transmitter:
      NodeName: ADAS
    signals:
      - name: SPEED_LIMIT
        multiplexer_indicator: Plain
        start_bit: 47
        size: 8
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: mph
        receivers:
          - XXX
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 59
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 61
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 43604
    name: LKAS_HUD_2
    size: 8
    transmitter:
      NodeName: ADAS
    signals:
      - name: COUNTER_2
        multiplexer_indicator: Plain
        start_bit: 7
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: SET_ME_X01
        multiplexer_indicator: Plain
        start_bit: 5
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: LKAS_BOH_1
        multiplexer_indicator: Plain
        start_bit: 15
        size: 6
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 63
        unit: ""
        receivers:
          - XXX
      - name: LEFT_LANE
        multiplexer_indicator: Plain
        start_bit: 23
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: RIGHT_LANE
        multiplexer_indicator: Plain
        start_bit: 21
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: LKAS_BOH_2
        multiplexer_indicator: Plain
        start_bit: 30
        size: 5
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 31
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 61
        size: 2
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 59
        size: 4
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
comments:
  - Plain:
      comment: IMPORT _honda_common.dbc
  - Plain:
      comment: IMPORT _bosch_2018.dbc
  - Plain:
      comment: IMPORT _lkas_hud_8byte.dbc
  - Plain:
      comment: IMPORT _bosch_standstill.dbc
  - Plain:
      comment: IMPORT _steering_sensors_a.dbc
  - Plain:
      comment: IMPORT _gearbox_common.dbc
  - Signal:
      message_id:
        Standard: 456
      name: IDLESTOP_ALLOW
      comment: allows car to turn off engine at a standstill
  - Signal:
      message_id:
        Standard: 456
      name: STANDSTILL
      comment: set to 1 when camera requests -4.0 m/s^2
  - Signal:
      message_id:
        Standard: 495
      name: SPEED_LIMIT
      comment: Defaults to 0xFF if no speed limit found
value_descriptions:
  - Signal:
      message_id:
        Standard: 477
      name: GEAR_MT
      value_descriptions:
        - id: 14
          description: reverse
        - id: 6
          description: 6th
        - id: 5
          description: 5th
        - id: 4
          description: 4th
        - id: 3
          description: 3rd
        - id: 2
          description: 2nd
        - id: 1
          description: 1st
        - id: 0
          description: Clutch
